### PR TITLE
docs: DLT-1399 add empty state section

### PR DESCRIPTION
**Pull request by fede-dp**
# DLT-1399 Add empty state section

## Obligatory GIF (super important!)

Obligatory GIF

## :hammer_and_wrench: Type Of Change

These types will increment the version number on release:

- [ ] Fix
- [ ] Feature
- [ ] Performance Improvement
- [ ] Refactor

These types will not increment the version number, but will still deploy to documentation site on release:

- [x] Documentation
- [ ] Build system
- [ ] CI
- [ ] Style (code style changes, not css changes)
- [ ] Test
- [ ] Other (chore)

## :book: Jira Ticket

[DLT-1399]

## :pencil: Checklist

For all PRs:

- [x] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [x] I have reviewed my changes.
- [x] I have added all relevant documentation.
- [x] I have considered the performance impact of my change.

## :link: Sources

[DLT-1399]: https://dialpad.atlassian.net/browse/DLT-1399?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by iropolo** [comment]
Empty state component doc preview: https://dialtone.dialpad.com/deploy-previews/pr-282/components/empty-states.html

File: apps/dialtone-documentation/docs/_data/site-nav.json
Comment (braddialpad) [inline apps/dialtone-documentation/docs/_data/site-nav.json]: Why are we removing this?

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: This is really small. I can't read the numbers.

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: To nest an unordered list under an ordered list I think you have to put 3 spaces of indentation before it (super weird markdown thing). Right now it looks kinda off.

Screenshot 2024-05-28 at 12 12 57 PM

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: Same problem here with the images being too small. Can't really even read the text.

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "gonna" is slang. Change to "going to"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "but not as much for them to feel overwhelmed." -> "but not so much they feel overwhelmed"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "always keep" -> "always give"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "Use elements too colorful to take away the attention from the main task" -> "Use elements so colorful they take away attention from the main task"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "AI"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: Will make it a bit more readable if you bold these prefixes ex "Titles:" "Body:"

File: apps/dialtone-documentation/docs/_data/site-nav.json
Comment (braddialpad) [inline apps/dialtone-documentation/docs/_data/site-nav.json]: state, rather than "states" to be consistent with all our other pages, and the actual page heading.

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: There are some rules to follow when using empty state.

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: decapitalize "If"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: Make these lists with `-` instead of `<li>`

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "There are multiple ways to use empty state as shown below:"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: hmm why do we need to set d-mt8 here?

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: Convert to markdown instead of HTML

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "Remember the below rules when setting the size of the empty state component"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "chages" -> "changes"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: "Below are some examples of how to render empty state at different sizes"

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-states.md]: At a minimum, empty state requires a body of content. It can optionally contain a header with buttons, or a footer with buttons/text.

**Review by braddialpad — CHANGES_REQUESTED** [review]
Thanks! Some fixes. Mostly small language things.

File: apps/dialtone-documentation/docs/_data/site-nav.json
Comment (iropolo) [reply apps/dialtone-documentation/docs/_data/site-nav.json]: No clue, seems like it was not intended

Comment (iropolo) [reply apps/dialtone-documentation/docs/_data/site-nav.json]: Now restored 👍🏼

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (iropolo) [reply apps/dialtone-documentation/docs/components/empty-states.md]: Image updated

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (iropolo) [reply apps/dialtone-documentation/docs/components/empty-states.md]: Updated

image

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (iropolo) [reply apps/dialtone-documentation/docs/components/empty-states.md]: I updated the images, if it still not convince we will had to wait for Fede to take a look

File: apps/dialtone-documentation/docs/components/empty-states.md
Comment (iropolo) [reply apps/dialtone-documentation/docs/components/empty-states.md]: removed all `d-mt8` seems like a leftover

**Comment by iropolo** [comment]
I will take care of the fixes.

Updated doc, @braddialpad thanks for your time doing this review, please `unresolve conversation` if there is sth left.
Thanks,

File: apps/dialtone-documentation/docs/components/empty-state.md
Comment (braddialpad) [inline apps/dialtone-documentation/docs/components/empty-state.md]: Hmmm these are not outputting correctly for some reason.. maybe needs a line break after the header?

Screenshot 2024-05-28 at 3 30 17 PM

Comment (braddialpad) [reply apps/dialtone-documentation/docs/components/empty-state.md]: Oh wait.. I think this just needs to be deleted. It is a repeat of whats under the `## Variable size` section.

Still can put a line break after `## Variants and examples`

**Review by braddialpad — APPROVED** [review]
Good to merge after fixing my one comment.

**Comment by francisrupert** [comment]
Since @fede-dp instantiated this PR, was off Monday and Tuesday, let's ensure he's had a chance to go over it before merging.

**Comment by iropolo** [comment]
> Since @fede-dp instantiated this PR, was off Monday and Tuesday, let's ensure he's had a chance to go over it before merging.

Okay... lets wait for Fede to check it out 👍🏼

**Comment by francisrupert** [comment]
* I'm seeing some subtle **spacing and color** differences between the Figma component and https://dialtone.dialpad.com/deploy-previews/pr-282/components/empty-state.html across the difference sizes. 
  * I'll let @fede-dp go over that carefully with you. 

* For the HTML/CSS, please defer almost completely to `dt-stack` to emulate the spacing and structure from the Figma file — in particular, avoid `margin`. You can think of Figma's autolayout structure as guidance for structuring that.

* Avoid CSS Utilities for official DT components, e.g. `class="d-p32"`, instead applying correct use of BEM css structure to style w/ the appropriate spacing CSS variable.

* Avoid nested selectors, e.g. `.d-empty-state--size-lg p`. Keep them flat. 

* For all text styles, use the corresponding [text style](https://dialtone.dialpad.com/design/typography/) rather than declaring it in CSS, e.g. `class="d-headline--lg"`, `class="d-body--md"`. Note that these are not strictly CSS Utilities.

* Depending on the flow and context of the Empty State usage, `<h1>` very likely won't hierarchically hold up. Unless you provide a way to force the consuming engineer to assign an `<h#>` to it, you're almost better off making it a `<div>` with the appropriate `class="d-headline--{SIZE}"`. 

* For the title and paragraph, use [`text-wrap: balance`](https://developer.chrome.com/docs/css-ui/css-text-wrap-balance). This will avoid awkward multi-line text renderings.

**Comment by fede-dp** [comment]
Hey Nacho! sorry for the late response, 
looking great overall! a few comments about spacing:

- we need to add `gap: var(--dt-space-450)` between the h1 and p 
preview:
image
figma:
image

i like the approach of removing the context in the images, will update that to keep using rounded corners so it's more similar to the window of the app/browser.

**Comment by francisrupert** [comment]
@fede-dp Colors: All foreground content (icons, text) appear to be using **primary** foreground color while the Figma file (unless I'm looking at the wrong Figma branch) has a nice balance of Primary, Secondary, Tertiary, and Muted foreground colors.

**Comment by fede-dp** [comment]
Regarding Colors:
- everything is in primary, adding image of the figma here 👇🏼 
image
image

**Comment by fede-dp** [comment]
> @fede-dp Colors: All foreground content (icons, text) appear to be using **primary** foreground color while the Figma file (unless I'm looking at the wrong Figma branch) has a nice balance of Primary, Secondary, Tertiary, and Muted foreground colors.

woops was typing this and saw that. Francis make it more clearer than i am. Thanks!

**Comment by iropolo** [comment]
Not related to doc, but I updated the component.
### **These changes should be in a different ticket/PR and not in the current one. (out of scope)**

Added the spacing between `headerText` and `bodyText`.
Seems like `headerText` already have `var(--dt-color-foreground-primary)`.
Updated color to `color: var(--color-foreground-secondary, #3A3A3A);` on `bodyText`.
Updated icon color to `color: var(--dt-color-foreground-muted)`.

Not sure where is applied a `Tertiary` color.


Edit: Didn't see https://github.com/dialpad/dialtone/pull/282#issuecomment-2136476788 , Im taking a look now.

**Comment by iropolo** [comment]
Update:

Moving to `dt-stack`, removed margin.
`emptyStateContentSpacing` seems like it was a leftover, removed `d-p` utilities.
`h1` -> `div`.
Moved from css to classes to handle dynamically headline and body text.
Added `text-wrap: balance;` property. It does not has great browser support but it will work on modern ones.

Doubts on 

```
Avoid nested selectors, e.g. .d-empty-state--size-lg p. Keep them flat.
```
What you mean with it? We have nested selectors all over the application, not sure about the special benefit to change it.


Lemme know if I understand your suggestions and the update fit with it.
Also if `dialtone-vue2` changes are ok I will move it to `-vue3`
And update tests accordingly.

**Comment by fede-dp** [comment]
i think the variant of the secondary is broken in dark mode 
image

**Comment by braddialpad** [comment]
> Doubts on
> 
> ```
> Avoid nested selectors, e.g. .d-empty-state--size-lg p. Keep them flat.
> ```
> 
> What you mean with it? We have nested selectors all over the application, not sure about the special benefit to change it.

Nesting creates higher specificity which can make things very confusing when trying to style elements directly. It's good practice to keep things as flat as possible for maximum scalability. See the following example for what I mean.

https://css-tricks.com/strategies-keeping-css-specificity-low/#aa-dont-avoid-classes-to-begin-with

**Comment by braddialpad** [comment]
What's the status of this one? are we waiting on something?

**Comment by fede-dp** [comment]
In the preview I still see the description text color isn't fixed
image
other than that I think the docs is great!

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: Avoid nested selectors, keep them flat. This renders as `.d-empty-state .d-icon {}`.

Change this selector to 
```less
&__icon {
  color: var(--dt-color-foreground-muted)
}
```

Wrap the `<dt-icon>`

```html
<span class="d-empty-state__icon">
  <dt-icon>
</span>
```

The same notion applies to the illustration...

```html
<span class="d-empty-state__illustration">
  (illustration)
</span>
```
...even though there isn't currently a style to apply to it.

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: `&--sm`

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: `&__content`

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: `&__header-text`

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: `&--md`

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: `color: var(--dt-color-foreground-secondary);`

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: `&__body-text`

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: Nested selector. This should be a _**variant**_ of `.d-empty-state__content`... move to there, such that it appears as...

```less
&__content {
  padding: var(--dt-space-600) 0;
}
&__content--sm {
  padding: var(--dt-space-500) 0;
}
```

And reflect that selector in corresponding Vue component.

File: apps/dialtone-documentation/docs/components/empty-state.md
Comment (francisrupert) [inline apps/dialtone-documentation/docs/components/empty-state.md]: I'm guessing all the `htmlCode` examples need to be updated to reflect updated HTML rendering.

**Review by francisrupert — CHANGES_REQUESTED** [review]
Almost there. This is an good example of a simple component to gain stronger familiarity with BEM. Note that changing based on these comments would need to be reflected elsewhere, e.g. Vue component, tests.

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [reply packages/dialtone-css/lib/build/less/components/empty-state.less]: `__` is for **child** elements, e.g. `.d-empty-state__content` 
* iow, "the content within `.d-empty-state`"

`--` is for **variants**, e.g. `.d-empty-state--sm`
* iow, "the small variant of `.d-empty-state`"

File: apps/dialtone-documentation/docs/components/empty-state.md
Comment (francisrupert) [inline apps/dialtone-documentation/docs/components/empty-state.md]: Shouldn't all the Vue `<code-well-header>` and `vueCode` examples use `<dt-stack>` and `<dt-btn>` rather than its raw HTML?

Comment (braddialpad) [reply apps/dialtone-documentation/docs/components/empty-state.md]: Yeah I would think so

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (iropolo) [reply packages/dialtone-css/lib/build/less/components/empty-state.less]: Added `line-height: 0` cos we are wrapping it in a `<span>` element.

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (iropolo) [reply packages/dialtone-css/lib/build/less/components/empty-state.less]: Updated classes in `empty_state_constants.js` to match this update

```
export const EMPTY_STATE_SIZE_MODIFIERS = {
  sm: 'd-empty-state--size--sm',
  md: 'd-empty-state--size--md',
  lg: 'd-empty-state--size--lg',
};
```

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (iropolo) [reply packages/dialtone-css/lib/build/less/components/empty-state.less]: Added a `contentClass` to handle this.

File: packages/dialtone-css/lib/build/less/components/empty-state.less
Comment (francisrupert) [inline packages/dialtone-css/lib/build/less/components/empty-state.less]: Let's change this to `var(--dt-color-foreground-tertiary)` for now. Unfortunately quite a few icons haven't been flattened into a single `<path>`, so the translucent `muted` color results in overlapping parts (attached image). 

I'll log an issue to identify and fix those.

image

**Review by francisrupert — APPROVED** [review]
Good quick fixes! Just one little update, and good to merge. I'm going to follow that up with an edit of the doc's contents for conciseness, grammar, and organization.

/c @fede-dp

**Pull request merged**
